src/care/LoopFuser.cpp

```diff
@@ -347,127 +231,98 @@ void LoopFuser::flush_parallel_scans() {
-      }
-#endif
-      int offset = actionIndex == 0 ? 0 : offsets[actionIndex-1];
-#ifdef FUSER_VERY_VERBOSE
-      if (verbose) {
-         printf("new offset %i\n", offset);
-         printf("LOOPFUSER::scan_var[%i] = %i\n", i, scan_var[index]-scan_var[offset]);
-      }
-#endif
-      conditionals[actionIndex](index,true,scan_var[index]-scan_var[offset],-1,-1);
-   } CARE_STREAM_LOOP_END
+   conditional_workgroup cw = m_conditionals.instantiate();
+   conditional_worksite cws = cw.run(scan_var.data(chai::GPU,true), offsets, end, fusible_registers{});
 
-   if (!async) {
-      // need to do a synchronize data so subsequent writes to this fuser's buffers do not overlap with zero copy reads.
-      // If async is on, programmer takes responsibility for ensuring this does not happen.
-      care::syncIfNeeded();
-   }
 
    scan_var.free();
+   if (verbose) {
+     printf("done with flush_counts_to_offsets_parallel_scans at %s:%i with %i,%i\n", fileName,lineNumber, m_action_count, m_max_action_length);
+   }
+   reset(async, fileName, lineNumber);
 }
 
-void LoopFuser::flushActions(bool async) {
-#ifdef FUSER_VERBOSE
-   printf("Loop fuser flushActions\n");
-#endif
+void LoopFuser::flushActions(bool async, const char * fileName, int lineNumber) {
+   if (verbose) {
+      printf("Loop fuser flushActions\n");
+   }
    if (m_action_count > 0) {
       if (m_is_scan) {
-#ifdef FUSER_VERBOSE
-         printf("loop fuser flush parallel scans\n");
-#endif
-         flush_parallel_scans();
+         if (verbose) {
+            printf("loop fuser flush parallel scans\n");
+         }
+         flush_parallel_scans(fileName, lineNumber);
       }
       else if (m_is_counts_to_offsets_scan) {
-#ifdef FUSER_VERBOSE
-         printf("loop fuser flush counts to offsets scans\n");
-#endif
-         flush_parallel_counts_to_offsets_scans(async);
+         if (verbose) {
+            printf("loop fuser flush counts to offsets scans\n");
+         }
+         flush_parallel_counts_to_offsets_scans(async, fileName, lineNumber);
       }
       else {
          if (m_preserve_action_order) {
-#ifdef FUSER_VERBOSE
-            printf("loop fuser flush order preserving actions\n");
-#endif
-            flush_order_preserving_actions(async);
+            if (verbose) {
+               printf("loop fuser flush order preserving actions\n");
+            }
+            flush_order_preserving_actions(async, fileName, lineNumber);
          }
          else {
-#ifdef FUSER_VERBOSE
-            printf("loop fuser flush parallel actions\n");
-#endif
-            flush_parallel_actions(async);
+            if (verbose) {
+               printf("loop fuser flush parallel actions\n");
+            }
+            flush_parallel_actions(async, fileName, lineNumber);
          }
       }
    }
    for (auto arr : m_to_be_freed) {
       arr.free();
    }
    m_to_be_freed.clear();
-   reset(async);
```

Comment:
I'm assuming a reset operation is no longer needed?

Reply:
yup, the reset got pushed into each of the flush commands to optimize the number of gpuDeviceSynchronize calls.